### ISSUE TITLE
[REVIEW] Tests are not asserting for expected functionality

### ISSUE THREAD

**Issue by santisoler** [underworldcode/stripy#29]
Part of https://github.com/openjournals/joss-reviews/issues/1410

One of the JOSS requirements is to have manual or automated tests to verify if the functionality of the package.
You have successfully added manual tests inside `stripy/stripy-src/tests` that check:
- if required dependencies and some other packages are already installed,
- if Cartesian and Spherical interpolation, derivative and smoothing don't rise errors,
- if extra methods of Cartesian and Spherical Triangulations don't rise errors.

So, most tests are only checking if the code doesn't rise errors when expected arguments are passed.
Nevertheless, they could be improved by asserting that the returned results are equal (or sufficiently close) to the expected ones.
I consider that the tests lack these kind of assertions, so I would **recommend implementing** some of them in order **to meet JOSS requirements**.

I'm willing to help you implementing these assertion tests, please don't hesitate to ask!

## Pytest

A great tool to perform tests is  [`pytest`](https://docs.pytest.org/en/latest/).
It's very easy to use and also let us automate the tests run.
(I've seen that you intend to run the [`test_imports.py`](https://github.com/underworldcode/stripy/blob/master/stripy-src/tests/test_imports.py) after installation through pip. This could be very easily implemented with `pytest`).

`pytest` allows you to run all the tests functions from a single line and show which ones are failing, for example:
```
pytest -v stripy
```

The basis of `pytest` is the `assert` statement. It raises an error in case that the boolean statement after `assert` is `False`.
So, `pytest` will show a certain test function as failed in case it raised some error, e.g. the `AssertionError`.
`pytest` also can catch errors where are expected to occur.

For example, consider a simple function to compute the hypotenuse of a right triangle given it sides:
```python
import numpy as np

def pythagorean(a, b):
    """
    Compute hypotenuse of a right triangle
    """
    if a <= 0 or b <= 0:
        raise ValueError(
            "Not valid values of sides lengths: {}, {}. ".format(a, b)
            + "They should be greater than zero."
        )
    return np.sqrt(a ** 2 + b ** 2)
```

We could write two test functions for it:
- one for asserting that the returned hypotenuse and
- another one that catches the `ValueError` if one of the sides is not greater than zero.
```python
import numpy as np
import numpy.testing as npt
from pytest import raises

from pythagorean import pythagorean

def test_pythagorean():
    "Check if hypotenuse is correctly computed"
    a, b = 1, 1
    npt.assert_allclose(pythagorean(a, b), np.sqrt(2))
    a, b = 2, 1
    npt.assert_allclose(pythagorean(a, b), np.sqrt(5))
    a, b = 1, 2
    npt.assert_allclose(pythagorean(a, b), np.sqrt(5))

def test_invalid_sides():
    "Check if pythagorean rises errors on invalid sides"
    with raises(ValueError):
        pythagorean(a=0, b=1)
    with raises(ValueError):
        pythagorean(a=-1, b=1)
    with raises(ValueError):
        pythagorean(a=1, b=0)
    with raises(ValueError):
        pythagorean(a=1, b=-1)
    with raises(ValueError):
        pythagorean(a=-1, b=-1)
```

In summary, the `pytest` implementation **is not required** for JOSS acceptance, but if you want to implement it, I'm willing to help!

**Comment by lmoresi** [comment]
Agreed - pytest is great. The current tests are a bit lengthy for the pytest approach but I have  put them into the pytest format. There is some scipy issue at the moment that needs to be sorted out

**Comment by santisoler** [comment]
:+1: Let me know if you need any help with those issues!

**Comment by brmather** [comment]
This commit ca6e6511cba6e3cb29167b1464a25e8703b2496a tests the following functionality of the spherical and Cartesian triangulations:

- Derivatives `mesh.gradient(...)`
- Nearest neighbour interpolation `mesh.interpolate_linear(...)`
- Linear interpolation `mesh.interpolate_linear(...)`
- Cubic interpolation `mesh.interpolate_cubic(...)`
- Smoothing `mesh.smoothing(...)`

All tests are passing with `pytest` (which is a very useful tool by the way!) Let me know if there are any other tests you can think of, otherwise we can close this issue.

Cheers :+1:

**Comment by lmoresi** [comment]
There will always be more tests - and we can check coverage with the relevant pytest plugin but the important thing is that we now have a reliable and extensible testing framework. We should run these in the final stages of the docker build to ensure we have at least some continuous checking. Then I would assert that we have done as asked and can close the issue.

Sent from my iPhone

On 29 May 2019, at 19:38, Ben Mather <notifications@github.com<mailto:notifications@github.com>> wrote:


This commit ca6e651<https://github.com/underworldcode/stripy/commit/ca6e6511cba6e3cb29167b1464a25e8703b2496a> tests the following functionality of the spherical and Cartesian triangulations:

  *   Derivatives mesh.gradient(...)
  *   Nearest neighbour interpolation mesh.interpolate_linear(...)
  *   Linear interpolation mesh.interpolate_linear(...)
  *   Cubic interpolation mesh.interpolate_cubic(...)
  *   Smoothing mesh.smoothing(...)

All tests are passing with pytest (which is a very useful tool by the way!) Let me know if there are any other tests you can think of, otherwise we can close this issue.

Cheers 👍

**Comment by brmather** [comment]
One can never have too many tests! Since 01defbdc81b383ce07306df0a378b953b21d588a `pytest` will run everything in the _tests_ directory.
